docs/auto-setup.adoc

```diff
@@ -0,0 +1,111 @@
+= Automatic setup on virtual machine
+
+== Prerequisites
+
+You need to install:
+
+- https://www.packer.io/downloads[Packer]
+- https://www.vagrantup.com/downloads[Vagrant]
+- https://www.virtualbox.org/wiki/Downloads[VirtualBox]
+
+[TIP]
+VirtualBox installation may fail on macOS Catalina. If it's your case, please look
+at https://apple.stackexchange.com/questions/372492/virtualbox-installation-failed-on-macos-catalina[this issue].
+
+== Installing the system
+
+First, go to the `deployments/local-setup-environment` directory and build the environment image:
+```
+packer build -only=vagrant packerfile.json
+```
+An output directory named `vagrant-box` containing `package.box` file should be created.
+
+Then, switch the directory to `deployments/local-setup-instance` and import
+the image:
+```
+vagrant box add --force --name=local-setup-environment ./../local-setup-environment/vagrant-box/package.box
+```
+Finally, you can run the virtual machine by doing:
+```
+vagrant up
+```
+A new virtual machine will be created and provisioned accordingly. When
+this command terminates successfully, the machine should be ready to work
+and have all the auxiliary software like Geth, Bitcoin Core, ElectrumX
+and Core/ECDSA clients running.
+
+== How to interact with the system
+
+You can interact with the system through the tBTC dApp or automated
+end-to-end tests. Before you start interacting, you have to connect
+to the virtual machine. First, switch to the `deployments/local-setup-instance`
+directory and do:
+```
+vagrant ssh
+```
+
+You will be moved to the virtual machine home directory where you can find
+the `local-setup` directory where all installed things live.
+
+=== Keep Dashboard dApp
+
+To run the Keep Dashboard dApp invoke:
+```
+./run-keep-dashboard.sh
+```
+
+=== tBTC dApp
+
+To run the tBTC dApp against the local Bitcoin network,
+make sure the tBTC dApp is configured to work with the local ElectrumX instance.
+You can configure this in `./tbtc-dapp/src/wrappers/web3.js` file,
+where `TBTC.withConfig` line occurs.
+Then, you can invoke:
+```
+./run-tbtc-dapp.sh
+```
+The application will be available on `http://localhost:3000`.
+
+=== E2E tests
+
+To run the automated end-to-end scenario switch to Node 14.3.0:
+```
+nvm use 14.3.0
+```
+Then invoke:
+```
+./run-e2e-test.sh
```

Comment:
It fails with:
```
vagrant@ubuntu-xenial:~/local-setup$ nvm use 14.3.0
manpath: can't set the locale; make sure $LC_* and $LANG are correct
Now using node v14.3.0 (npm v6.14.5)
vagrant@ubuntu-xenial:~/local-setup$ ./run-e2e-test.sh
npm WARN e2e@1.0.0 No description
npm WARN e2e@1.0.0 No repository field.
npm WARN e2e@1.0.0 No license field.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: semver@6.3.0 (node_modules/@keep-network/tbtc.js/node_modules/@keep-network/keep-ecdsa/node_modules/@web3-js/scrypt-shim/node_modules/semver):
npm WARN enoent SKIPPING OPTIONAL DEPENDENCY: ENOENT: no such file or directory, open '/home/vagrant/local-setup/e2e/node_modules/@keep-network/tbtc.js/node_modules/@keep-network/keep-ecdsa/node_modules/@web3-js/scrypt-shim/node_modules/semver/package.json.2760576506'

npm ERR! code ENOENT
npm ERR! syscall open
npm ERR! path /home/vagrant/local-setup/e2e/node_modules/@keep-network/tbtc.js/node_modules/@keep-network/keep-ecdsa/node_modules/@openzeppelin/contracts-ethereum-package/package.json.2488701596
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, open '/home/vagrant/local-setup/e2e/node_modules/@keep-network/tbtc.js/node_modules/@keep-network/keep-ecdsa/node_modules/@openzeppelin/contracts-ethereum-package/package.json.2488701596'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/vagrant/.npm/_logs/2020-07-22T12_13_13_015Z-debug.log
```

Reply:
Hmm, strange. It works on my machine. Can you provide `git submodule` output here?

Reply:
```
➜  local-setup git:(local-setup-vm) ✗ git submodule
-c6af1a5c49860b7f1ba2f015adf820c01aa7b2e7 keep-core
-89bd7bffea7282da3212b9a3f6b4f559725c3e67 keep-ecdsa
-efea3a8590e821c617256d6ef56ada956319c6ba relays
-d82b54abe41d6d96c01734c9a5d6926b47447b20 tbtc
-e00534f812c7e4b54f27f5b8f41bb79f6f7c947f tbtc-dapp
-6890d466d1a32793d4834a1e0d35d30c0acd047e tbtc.js
```

Reply:
This one is from the VM's `local-setup` directory?

Reply:
Nope, it's from my machine in this repository, not from the VM but the VM says the same:
```
vagrant@ubuntu-xenial:~/local-setup$ git submodule
 c6af1a5c49860b7f1ba2f015adf820c01aa7b2e7 keep-core (v1.2.4-rc-291-gc6af1a5)
 89bd7bffea7282da3212b9a3f6b4f559725c3e67 keep-ecdsa (v1.1.0-142-g89bd7bf)
 efea3a8590e821c617256d6ef56ada956319c6ba relays (v0.0.1-alpha-93-gefea3a8)
 d82b54abe41d6d96c01734c9a5d6926b47447b20 tbtc (v0.14.0-rc-393-gd82b54a)
 e00534f812c7e4b54f27f5b8f41bb79f6f7c947f tbtc-dapp (v0.14.0-rc-70-ge00534f)
 6890d466d1a32793d4834a1e0d35d30c0acd047e tbtc.js (v0.12.0-rc-50-g6890d46)
```

Reply:
So you have old submodules on the VM `local-setup` somehow. This is probably the cause of failing tests. Pretty strange, because your `install-repositories.sh` has:
```
git submodule update --init --recursive --remote --rebase --force 
```
which should take the recent versions. I'll try to reproduce the problem on my machine.

Reply:
Yep, after running `install-repositories.sh` submodules get updated but it seems like it hasn't been triggered for vagrant setup for some reason.

Reply:
I've not reproduced this on my machine so far. Maybe this was a one-off glitch. If you observe it again, please ping me with as many details as you can.

Reply:
That's weird - I tried two times and I had correct git submodules downloaded. And test is passing :tada:

```
Initial TBTC balance for account 0xD6b0a1CA8f0641B97eFeC0F1eD73d72e58b38FA5 is: 0

Starting deposit number [1]...

Creating new deposit contract with lot size 100000 satoshis...
Looking up new deposit with address 0x62FEd5bdEb1AD7D9A673fBB9a0ED971a4d26810C backed by keep at address 0x715AB49Cfb48eBa2997aa611BC260D8e5E19103c...
Monitoring deposit 0x62FEd5bdEb1AD7D9A673fBB9a0ED971a4d26810C for transition to ACTIVE.
Waiting for deposit 0x62FEd5bdEb1AD7D9A673fBB9a0ED971a4d26810C keep public key...
Waiting for deposit 0x62FEd5bdEb1AD7D9A673fBB9a0ED971a4d26810C to retrieve public key...
Found public key for deposit 0x62FEd5bdEb1AD7D9A673fBB9a0ED971a4d26810C...
Monitoring Bitcoin for transaction to address bcrt1qa2cenxr3z62x3lujdvzg8quhtk9vhdeqwk032z...
Connecting to electrum server...
	Got deposit address: bcrt1qa2cenxr3z62x3lujdvzg8quhtk9vhdeqwk032z ; fund with: 100000 satoshis please.
Now monitoring for deposit transaction...
You are connected to an ElectrumX 1.14.0 server.
Negotiated version: [ElectrumX 1.14.0,1.4.2]
Deposit transaction sent
Received notification for script hash: [ca4cc0476e0dfb811aa0ef4701a88ded6e0b6b58759699c5cb5fe029c8b44502] with status: [b8b978a474a6e3ec7eb0546506d9ebcd6058aff5613c6b5e238c6db6ae009b1b]
Closing connection to electrum server...
Waiting for 1 confirmations for Bitcoin transaction 789977615255e567942e512792955aa331e899da3c4aebdad220e38080b94703...
Connecting to electrum server...
You are connected to an ElectrumX 1.14.0 server.
Negotiated version: [ElectrumX 1.14.0,1.4.2]
Registered listener for blockchain.headers.subscribe event
Received notification of a new block at height: [1142]
Closing connection to electrum server...
Submitting funding proof to deposit 0x62FEd5bdEb1AD7D9A673fBB9a0ED971a4d26810C for Bitcoin transaction 789977615255e567942e512792955aa331e899da3c4aebdad220e38080b94703...
Connecting to electrum server...
You are connected to an ElectrumX 1.14.0 server.
Negotiated version: [ElectrumX 1.14.0,1.4.2]
Closing connection to electrum server...
Deposit 0x62FEd5bdEb1AD7D9A673fBB9a0ED971a4d26810C transitioned to ACTIVE.
Deposit is active, minting...
Approving transfer of deposit 0x62FEd5bdEb1AD7D9A673fBB9a0ED971a4d26810C TDT to Vending Machine...
Minting TBTC...
Found Transfer event for 999500000000000 TBTC.

Deposit 0x62FEd5bdEb1AD7D9A673fBB9a0ED971a4d26810C has been created successfully.

Starting deposit number [2]...

Creating new deposit contract with lot size 100000 satoshis...
Looking up new deposit with address 0xe184795593948c584F23Ee529E6022C750f875ea backed by keep at address 0x4c23B2ad526F2ebA15800f37eF51E176bdDC4F22...
Monitoring deposit 0xe184795593948c584F23Ee529E6022C750f875ea for transition to ACTIVE.
Waiting for deposit 0xe184795593948c584F23Ee529E6022C750f875ea keep public key...
Waiting for deposit 0xe184795593948c584F23Ee529E6022C750f875ea to retrieve public key...
Found public key for deposit 0xe184795593948c584F23Ee529E6022C750f875ea...
Monitoring Bitcoin for transaction to address bcrt1qsfndd8frj2gps88fcerd0mgfgtz7mu664thfx7...
Connecting to electrum server...
	Got deposit address: bcrt1qsfndd8frj2gps88fcerd0mgfgtz7mu664thfx7 ; fund with: 100000 satoshis please.
Now monitoring for deposit transaction...
You are connected to an ElectrumX 1.14.0 server.
Negotiated version: [ElectrumX 1.14.0,1.4.2]
Deposit transaction sent
Received notification for script hash: [95c867b430ace33a04c0b1d6460927e010652c51a7ae48e20bdcfab4e574e282] with status: [8e3b48f10d7d38d41a69fd2a6594fe89c07f593400c4cc5c76283ef91d0933a3]
Closing connection to electrum server...
Waiting for 1 confirmations for Bitcoin transaction d15b9fc2247de6fe311dcfa985f26c94973aa25ca81f9b11d6544fcaae3b0ca1...
Connecting to electrum server...
You are connected to an ElectrumX 1.14.0 server.
Negotiated version: [ElectrumX 1.14.0,1.4.2]
Registered listener for blockchain.headers.subscribe event
Received notification of a new block at height: [1145]
Closing connection to electrum server...
Submitting funding proof to deposit 0xe184795593948c584F23Ee529E6022C750f875ea for Bitcoin transaction d15b9fc2247de6fe311dcfa985f26c94973aa25ca81f9b11d6544fcaae3b0ca1...
Connecting to electrum server...
You are connected to an ElectrumX 1.14.0 server.
Negotiated version: [ElectrumX 1.14.0,1.4.2]
Closing connection to electrum server...
Deposit 0xe184795593948c584F23Ee529E6022C750f875ea transitioned to ACTIVE.
Deposit is active, minting...
Approving transfer of deposit 0xe184795593948c584F23Ee529E6022C750f875ea TDT to Vending Machine...
Minting TBTC...
Found Transfer event for 999500000000000 TBTC.

Deposit 0xe184795593948c584F23Ee529E6022C750f875ea has been created successfully.
TBTC balance for account 0xD6b0a1CA8f0641B97eFeC0F1eD73d72e58b38FA5 after performing deposits should be: 1999000000000000. Checking assertion...

Starting redemption of the first deposit...

Generated reedemer address: bcrt1q744rnewv7ly4pxjym4jtkhhe36axyaps9dr23x
Looking up Deposit contract at address 0x62FEd5bdEb1AD7D9A673fBB9a0ED971a4d26810C...
Looking up Created event for deposit 0x62FEd5bdEb1AD7D9A673fBB9a0ED971a4d26810C...
Found keep address 0x715AB49Cfb48eBa2997aa611BC260D8e5E19103c.
Found existing Bitcoin address for deposit 0x62FEd5bdEb1AD7D9A673fBB9a0ED971a4d26810C...
Looking up UTXO size and transaction fee for redemption transaction...
Approving transfer of 1000000000000000 to the vending machine....
Initiating redemption of deposit 0x62FEd5bdEb1AD7D9A673fBB9a0ED971a4d26810C from vending machine...
Looking up latest redemption details for deposit 0x62FEd5bdEb1AD7D9A673fBB9a0ED971a4d26810C...
Finding or waiting for transaction signature for deposit 0x62FEd5bdEb1AD7D9A673fBB9a0ED971a4d26810C...
Looking for existing signed redemption transaction on Bitcoin chain for deposit 0x62FEd5bdEb1AD7D9A673fBB9a0ED971a4d26810C...
Connecting to electrum server...
You are connected to an ElectrumX 1.14.0 server.
Negotiated version: [ElectrumX 1.14.0,1.4.2]
Closing connection to electrum server...
Broadcasting signed redemption transaction to Bitcoin chain for deposit 0x62FEd5bdEb1AD7D9A673fBB9a0ED971a4d26810C...
Connecting to electrum server...
You are connected to an ElectrumX 1.14.0 server.
Negotiated version: [ElectrumX 1.14.0,1.4.2]
Closing connection to electrum server...
Waiting for 1 confirmations for Bitcoin transaction 9d482a622d28c3f4d03d592a4ad10daaf807c726e1dcc1d53abb887d171e1312...
Connecting to electrum server...
You are connected to an ElectrumX 1.14.0 server.
Negotiated version: [ElectrumX 1.14.0,1.4.2]
Registered listener for blockchain.headers.subscribe event
Received notification of a new block at height: [1151]
Closing connection to electrum server...
Transaction is sufficiently confirmed; submitting redemption proof to deposit 0x62FEd5bdEb1AD7D9A673fBB9a0ED971a4d26810C...
Connecting to electrum server...
You are connected to an ElectrumX 1.14.0 server.
Negotiated version: [ElectrumX 1.14.0,1.4.2]
Closing connection to electrum server...


Redemption outcome: Redeemed deposit 0x62FEd5bdEb1AD7D9A673fBB9a0ED971a4d26810C with Bitcoin transaction 9d482a622d28c3f4d03d592a4ad10daaf807c726e1dcc1d53abb887d171e1312.

TBTC balance for account 0xD6b0a1CA8f0641B97eFeC0F1eD73d72e58b38FA5 after performing redemption should be: 999000000000000. Checking assertion...
Received BTC amount for redeemer address bcrt1q744rnewv7ly4pxjym4jtkhhe36axyaps9dr23x after performing redemption should be: 0.0009985. Checking assertion...
Test completed successfully
```